### PR TITLE
The Return of the SettingsGateway Rewrite

### PR DESCRIPTION
**Pull request by kyranet**
### Description of the PR

Fixes https://github.com/dirigeants/klasa/issues/425.

### Changes Proposed in this Pull Request (List new items in CHANGELOG.MD)

**Added**:

- [[#525](https://github.com/dirigeants/klasa/pull/525)] Add GatewayStorage#sync so they can be registered. (kyranet)
- [[#487](https://github.com/dirigeants/klasa/pull/487)] Added Settings#resolve(...paths) (UnseenFaith)
- [[#477](https://github.com/dirigeants/klasa/pull/477)] Added core `KlasaClient#settingsDelete` event handler to destroy instances in other shards. (kyranet)
- [[#475](https://github.com/dirigeants/klasa/pull/475)] Added `KlasaClient#settingsSync` event. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Added `Gateway#create` and `Gateway#acquire`. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Added `SETTING_GATEWAY_CHOOSE_KEY` and `SETTING_GATEWAY_UNCONFIGURABLE_FOLDER` i18n keys into en-US. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Added `SettingsFolder`, with the common operations (update, reset...) for better OOP compliancy. (kyranet)
- [[#426](https://github.com/dirigeants/klasa/pull/426)] Added `Schema#configurableValues`, similar to `Schema#configurableKeys`. (kyranet)
- [[#426](https://github.com/dirigeants/klasa/pull/426)] Added `SettingsUpdateOptions.rejectOnError` to make `Settings#update` reject when any of the keys are invalid or the parse turns out wrong. (kyranet)

**Changed**:

- [[#828](https://github.com/dirigeants/klasa/pull/828)] Moved all SettingsGateway related options into a namespaced `settings` option. (UnseenFaith)
- [[#814](https://github.com/dirigeants/klasa/pull/814)] Changed `Settings#get` to always deep clone values. (UnseenFaith)
- [[#807](https://github.com/dirigeants/klasa/pull/807)] Changed `Language#get`'s generic types to include an array of function values. (Favna)
- [[#698](https://github.com/dirigeants/klasa/pull/698)] Rewritten SettingsGateway's guides. (UnseenFaith)
- [[#700](https://github.com/dirigeants/klasa/pull/700)] Changed the API for `SettingsFolder#{pluck,resolve}` to return arrays instead of objects. (Skillz4Killz)
- [[#814](https://github.com/dirigeants/klasa/pull/814)] deepClone objects and arrays in `Settings#get()`. (Faith)
- [[#545](https://github.com/dirigeants/klasa/pull/545)] Changed the entire API for `QueryBuilder` to extend a Map instead of being a pseudo-map. (kyranet)
- [[#480](https://github.com/dirigeants/klasa/pull/480)] Changed the entire mechanism for `GatewayDriver#register` to take a `GatewayStorage` instance. (kyranet)
- [[#475](https://github.com/dirigeants/klasa/pull/475)] Renamed `KlasaClient#{settingsUpdateEntry,settingsDeleteEntry,settingsCreateEntry}` to `KlasaClient#{settingsUpdate,settingsDelete,settingsCreate}`. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Modified `Schema#defaults` type from Object literal to a `SettingsFolder` instance. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Modified `Schema#defaults` to be a property instead of a getter. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Modified `Gateway#get` to take only id. For get or create, use `Gateway#acquire` instead. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Modified `Settings#{update,reset}` output to return `{ key: string, value: any, piece: Schema | SchemaPiece }` instead of `{ data: string, any], piece: Schema | SchemaPiece }` in the `updated` field. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Modified `Settings` completely, they're not longer dictionaries but (nested) `SettingsFolder`s. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Modified `Util.resolveGuild` to also resolve members. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Modified initialization order to prepare `Schedule` before initializing pieces. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Modified Schema#get overload to not check safely to reduce overhead. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Modified Settings#sync to force sync if the entry was not synchronized. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Renamed `Gateway#syncQueue` to `Gateway#syncMap`, changing also its type from `Collection<string, Promise<Settings>>` to `WeakMap<string,-  Promise<Settings>>`. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Renamed `GatewayDriver#type` to `GatewayDriver#name`. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Renamed `SettingsUpdateOptions.{action,arrayPosition,avoidUnconfigurable}` to `SettingsUpdateOptions.{arrayAction,arrayIndex,onlyConfigurable}`. (kyranet)
- [[#426](https://github.com/dirigeants/klasa/pull/426)] Made schema resolving much faster and reliable. (kyranet)
- [[#426](https://github.com/dirigeants/klasa/pull/426)] **[BREAKING]** `Settings#{list,resolveString}` -> `Settings#display`. (kyranet)
- [[#426](https://github.com/dirigeants/klasa/pull/426)] Tweaked conf and userconf commands to use restString for values. (kyranet)
- [[#426](https://github.com/dirigeants/klasa/pull/426)] Guild Settings will not longer need guild. (kyranet)
- [[#426](https://github.com/dirigeants/klasa/pull/426)] **[BREAKING]** Moved the guild parameter in Settings#update to `SettingsUpdateOptions`. (kyranet)
- [[#467](https://github.com/dirigeants/klasa/pull/467)] **[BREAKING]** Changed `GatewayDriver` to extend `Collection<string, Gateway>` as opposed to being a dictionary type. (kyranet)

**Removed**:

- [[#480](https://github.com/dirigeants/klasa/pull/480)] Removed `Gateway#store`. (kyranet)
- [[#471](https://github.com/dirigeants/klasa/pull/471)] Removed `Gateway#defaults`. Refer to `Schema#defaults` instead. (kyranet)
- [[#426](https://github.com/dirigeants/klasa/pull/426)] Removed `Gateway#getPath`. They're now resolved inside Settings. (kyranet)

**Fixed**:

- [[#823](https://github.com/dirigeants/klasa/pull/823)] Fixed `KlasaUser#toJSON`. (PyroTechniac)
- [[#819](https://github.com/dirigeants/klasa/pull/819)] Fixed return type of `Store.walk`. (PyroTechniac)
- [[#818](https://github.com/dirigeants/klasa/pull/818)] Fixed `userconf` command using the guilds gateway instead of the users one. (UnseenFaith)
- [[#817](https://github.com/dirigeants/klasa/pull/817)] Fixed sharding broadcast evals. (UnseenFaith)
- [[#487](https://github.com/dirigeants/klasa/pull/487)] Fixed role deserializer. (UnseenFaith)
- [[#426](https://github.com/dirigeants/klasa/pull/426)] Fixed typings for QueryBuilder. (kyranet)

### Semver Classification

- [ ] This PR only includes documentation or non-code changes.
- [ ] This PR fixes a bug and does not change the (intended) framework interface.
- [x] This PR adds methods or properties to the framework interface.
- [x] This PR removes or renames methods or properties in the framework interface.

### CODE REVIEW

**Comment by cfanoulis** [comment]
quick question what's the status on this?

**Comment by kyranet** [comment]
Pending on @UnseenFaith to add the last feature.

File: guides/Building Your Bot/CreatingPointsSystems.md [hunk 1]
```diff
@@ -33,7 +33,7 @@ module.exports = class extends Monitor {
 		if (!message.guild) return;
 
 		// Update the user's configuration entry by adding 1 to it.
-		await message.author.settings.update('experience', message.author.settings.experience + 1);
+		await message.author.settings.update('experience', message.author.get('experience') + 1);
```
Comment (tjrgg) [inline guides/Building Your Bot/CreatingPointsSystems.md]: Is that suppose to be `message.author.settings.get('experience')`?

Comment (kyranet) [reply guides/Building Your Bot/CreatingPointsSystems.md]: Whoops, yes, can you update your message to show a suggestion? So I can commit the fix co-authored-by you

Comment (tjrgg) [reply guides/Building Your Bot/CreatingPointsSystems.md]: Done.

**Comment by bdistin** [comment]
this has been done with the latest master rewrite, closing this PR, but keeping the branch around until we are sure we can depend on the klasa 0.6.0+

**Pull request closed without merging**